lib/ood_core/job/adapters/htcondor.rb

```diff
@@ -0,0 +1,473 @@
+                        args = args.map(&:to_s)
+                        script_args = script_args.map(&:to_s).map { |s| s.to_s.gsub('"', "'") } # cannot do double 
+                        env = env.to_h.each_with_object({}) { |(k, v), h| h[k.to_s] = v.to_s }
+
+                        tempfile = Tempfile.new("htcondor_submit")
+                        tempfile.close
+                        path = tempfile.path
+                        tempfile.unlink # unlink the tempfile so it can be used by condor_submit
+
+                        call("bash", "-c", "cat > #{path}", stdin: script)
+                        output = call("condor_submit", *args, env: env, stdin: "arguments=#{path.split("/").last} #{script_args.join(" ")}\ntransfer_input_files=#{path}").strip
+
+                        match = output.match(/(cluster )?(\d+)/)
+                        raise Error, "Failed to parse job ID from output: #{output}" unless match
+                        match[2]
+
+                    end
+
+                    # Run the `condor_rm` command to remove a job
+                    # @param id [#to_s] the id of the job to remove
+                    # @raise [Error] if `condor_rm` command exited unsuccessfully
+                    def remove_job(id)
+                        call("condor_rm", id.to_s)
+                    rescue Error => e
+                        raise Error, "Failed to remove job #{id}: #{e.message}"
+                    end
+
+                    # Place a job on hold using `condor_hold`
+                    # @param id [#to_s] the id of the job to hold
+                    # @raise [Error] if `condor_hold` command exited unsuccessfully
+                    def hold_job(id)
+                        id = id.to_s
+                        call("condor_hold", id)
+                    rescue Error => e
+                        raise Error, "Failed to hold job #{id}: #{e.message}"
+                    end
+
+                    # Release a job from hold using `condor_release`
+                    # @param id [#to_s] the id of the job to release
+                    # @raise [Error] if `condor_release` command exited unsuccessfully
+                    def release_job(id)
+                        id = id.to_s
+                        call("condor_release", id)
+                    rescue Error => e
+                        raise Error, "Failed to release job #{id}: #{e.message}"
+                    end
+
+                    # Retrieve job information using `condor_q`
+                    # @param id [#to_s] the id of the job
+                    # @param owner [String] the owner(s) of the job
+                    # @raise [Error] if `condor_q` command exited unsuccessfully
+                    # @return [Array<Hash>] list of details for jobs
+                    def get_jobs(id: "", owner: nil)
+                        args = []
+                        args.concat ["-constraint", "ClusterId == #{id}"] unless id.to_s.empty?
```

Comment:
The `id` here is supposed to be the job id. (at least it is in the upper layers like in the `info` API). But I see here you're using `"ClusterId == #{id}"`.

Is that a mistake or is that how htcondor uses job ids?

Reply:
`ClusterId` == top level ID of the job, where a cluster is an aggregate of jobs (i.e. array jobs)
`ProcId` would be the sub id: one can have 111002.123 as job id, where 111002 is the ClusterId and 123 the ProcId.
So far, this adapter does not deal with array jobs beyond that.